### PR TITLE
Load Main-map.png

### PR DESCRIPTION
**Pull request by Incompleteusern**
Looking for feedback.

### CODE REVIEW

File: common/src/main/java/com/wynntils/core/webapi/WebManager.java [hunk 1]
```diff
@@ -231,12 +237,89 @@ public static boolean tryLoadItemList() {
...
+    public static CompletableFuture<Boolean> tryLoadMap() {
+        if (apiUrls == null || !apiUrls.hasKey("MainMap")) return CompletableFuture.completedFuture(false);
+
+        final File mapDirectory = new File(API_CACHE_ROOT, "map");
+
+        mapDirectory.mkdirs();
+
+        final CompletableFuture<Boolean> result = new CompletableFuture<>();
+
+        handler.addAndDispatchAsync(new RequestBuilder(apiUrls.get("MainMap"), "main_map.info")
```
Comment (magicus) [inline common/src/main/java/com/wynntils/core/webapi/WebManager.java]: I am very, very skeptical towards bringing this old bad `CenterX` thing into Artemis.

I want to have a new, json-based metadata file with map information. I propose this:
`maps.json`:
```
[
{ "file": "main-map.png",
  "x1": -3443,
  "z1": 1002,
  "x2": 2443,
  "z2": -3922
} 
]
```

This is interpreted that one tile of the map (in this case, the entire world map) resides in the file `main-map.png`. The top left pixel on the map png corresponds to the block at the position X=-3443,Z=1002, and the bottom right corresponds to X=2443,Z=-3922. It is a failure if the size of the png does not match the size spanned by the coordinates 2D BB, as given in the metadata.

Comment (HighCrit) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: @magicus would we want to limit ourselves to 1 pixel = 1 block? We could have an option for higher resolution maps in the future. Also, should we add hashes to the metadata? That way we only have to download new/updated areas. Although that might prevent users from using a custom map.

Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: I agree, but this is not the pr to change things as this would likely require changes to Artemis. `CenterX` and `CenterZ` actually refer to the bottom right X and Z coordinates so they have been renamed here.

Comment (magicus) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: You mean that it would require changes to Athena? 

I still think this *is* the place to introduce such changes. It's better that you make a proper json file, even if you have to check it in as part of the resources of Artemis, than to write code that parses to old and bad stuff.

I also think that we make the Athena/server part faaaar to complicated. For static data caching, we just need to fetch it off a github repo. Wham, bam, done. The Athena logic is really only needed for stuff like user tracking, etc.

So better yet, make a json file and put it at [Data-Storage](https://github.com/Wynntils/Data-Storage) in e.g. a new `Artemis` directory, and get it using http from a raw link like `https://raw.githubusercontent.com/Wynntils/Data-Storage/master/item-changes.json`.

Comment (magicus) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: And yes, @HighCrit's idea of a hash is an excellent idea. If we download the metadata, and the hash matches the png file we have on disk in our cache, we're good, otherwise we need to refetch it.

Comment (magicus) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: @HighCrit Acutally, my first attempt at the json format included a `"scale": 1`, but then I removed it. I think we better start with that assumption. If we ever want to change it, it is easy to add the scale factor back. But in the meantime, we will be creating a lot of extra work for ourselves, trying to pass around a scale factor that is not used. And in the end, if it is not tested, it will likely be broken anyway the day we introduce a non-1-to-1 map. So my recommendation is: "Yes, it might be a good idea, but not just yet".

Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: The current implementation of map fetching on the serverside resides in [WynntilsWebsite-API](https://github.com/Wynntils/WynntilsWebsite-API/tree/master/maps).

Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: @magicus This should be added now, with the additional pr of https://github.com/Wynntils/WynntilsWebsite-API/pull/7.

File: common/src/main/java/com/wynntils/core/webapi/WebManager.java [hunk 2]
```diff
@@ -231,12 +234,69 @@ public static boolean tryLoadItemList() {
...
+    public static CompletableFuture<Boolean> tryLoadMap() {
+        if (apiUrls == null || !apiUrls.hasKey("AMainMap")) return CompletableFuture.completedFuture(false);
+
+        final File mapDirectory = new File(API_CACHE_ROOT, "map");
+
+        mapDirectory.mkdirs();
+
+        final CompletableFuture<Boolean> result = new CompletableFuture<>();
+
+        handler.addAndDispatch(new RequestBuilder(apiUrls.get("AMainMap"), "main_map.info")
```
Comment (magicus) [inline common/src/main/java/com/wynntils/core/webapi/WebManager.java]:
```suggestion
        handler.addAndDispatch(new RequestBuilder(apiUrls.get("AMainMap"), "map")
```

This is, afaict, just a marker that we use to distinguish different request. Better make it a bit nicer.

File: common/src/main/java/com/wynntils/core/webapi/WebManager.java [hunk 3]
```diff
@@ -231,12 +234,69 @@ public static boolean tryLoadItemList() {
...
+                .cacheTo(new File(mapDirectory, "main-map.txt"))
```
Comment (magicus) [inline common/src/main/java/com/wynntils/core/webapi/WebManager.java]:
```suggestion
                .cacheTo(new File(mapDirectory, "maps.json"))
```

File: common/src/main/java/com/wynntils/core/webapi/WebManager.java [hunk 4]
```diff
@@ -231,12 +234,69 @@ public static boolean tryLoadItemList() {
...
+                .handleJson(json -> {
+                    JsonObject mapData = json.getAsJsonArray().get(0).getAsJsonObject();
```
Comment (magicus) [inline common/src/main/java/com/wynntils/core/webapi/WebManager.java]: Since we only have a single map image, this works fine for now. But maybe add a comment describing this limitation?

File: common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        this.texture = texture;
+        this.x1 = x1;
+        this.z1 = z1;
+        this.x2 = x2;
+        this.z2 = z2;
+        this.textureWidth = textureWidth;
+        this.textureHeight = textureHeight;
+
+        // Remove this if we ever have non 1 to 1 maps
+        assert x2 - x1 == textureWidth;
```
Comment (magicus) [inline common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java:39]: We can calculate width and height from the input coordinates. 

However, we need to validate that this information matches the actual texture. So you should check the calculated width/height (from x12/z12) with `texture.getPixels().getWidth()` etc. And if it fails, throw an exception (a runtime exception is fine, you don't have to declare it).

File: common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    public int getWorldXPosition(double textureX) {
+        return (int) Math.round(textureX + x1);
+    }
+
+    public int getWorldZPosition(double textureY) {
+        return (int) Math.round(textureY + z1);
+    }
+
+    public ResourceLocation resource() {
+        if (mapResource == null) {
```
Comment (magicus) [inline common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java]: Why is this not done in the constructor?

File: common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java [hunk 7]
```diff
@@ -42,7 +42,7 @@ public ServerIcon(ServerData server, boolean allowStale, Consumer<ServerIcon> on
 
         // Try default
         ResourceLocation destination =
-                new ResourceLocation("wynntils/servers/" + Hashing.sha1().hashUnencodedChars(server.ip) + "/icon");
```
Comment (magicus) [inline common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java:45]: This seem unrelated..?

File: common/src/main/java/com/wynntils/core/webapi/WebManager.java [hunk 8]
```diff
@@ -67,6 +72,8 @@ public class WebManager {
 
     private static @Nullable WynntilsAccount account = null;
 
+    private static @Nullable MapProfile map = null;
```
Comment (magicus) [inline common/src/main/java/com/wynntils/core/webapi/WebManager.java]: Ditch the `@Nullable`. They are just making the code cluttered.

File: common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java (hunk 5 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java:39]: How does this differ from what occurs right now?

File: common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java (hunk 7 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java:45]: It's shoehorned in, as a small fix.

File: common/src/main/java/com/wynntils/core/webapi/WebManager.java (hunk 4 above)
Comment (HighCrit) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: We should also have variable names for the files. I commented on https://github.com/Wynntils/WynntilsWebsite-API/pull/7 with an idea on how to do that. Which is trivial, really.

File: common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java (hunk 6 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java]: I said more about this in the discord, but `TextureManager` is null when the mod initalizer for fabric is called, thus creating it in the constructor leads to a crash or an unregistered resource location.

File: common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java (hunk 7 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java:45]: What is the fix about? The texture does look fine to me?

File: common/src/main/java/com/wynntils/core/webapi/WebManager.java (hunk 4 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: The "obvious" solution is to name the files after the hash, I guess. But then we should really have some method of cleaning out old files as well. So if the files are named individually per segment, we need not bother about that part.

Comment (HighCrit) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: We could also remove all files that did not match any hash in that directory

File: common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java (hunk 5 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java:39]: 1) You need to pass in the height to the constructur, even though it can be calculated. Instead you have an assert to verify that the calculation is correct.

2) You need to check the actual texture dimensions.

Comment (magicus) [reply common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java:39]: Okay, I see now that you send in the texture size. But that's putting the responsibility on the caller. This should be done by this class instead.

File: common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java (hunk 6 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java]: Does the crash happen with the ResourceLocation constructor, or the `register` call? If it is the latter, the ResourceLocation should still be created in the constructor.

File: common/src/main/java/com/wynntils/core/webapi/WebManager.java [hunk 9]
```diff
@@ -231,12 +234,69 @@ public static boolean tryLoadItemList() {
                 .useCacheAsBackup()
                 .build());
 
-        handler.dispatch(false);
-
         // Check for success
         return isItemListLoaded();
     }
 
+    public static CompletableFuture<Boolean> tryLoadMap() {
```
Comment (magicus) [inline common/src/main/java/com/wynntils/core/webapi/WebManager.java]: Actually, I think we should try to make clear that we are handling more than one map, so the method should be like `tryLoadMaps`, the cache should be like:
```
final File mapsDirectory = new File(API_CACHE_ROOT, "maps");
```

etc.

And the more I think about it, the stronger I feel that we should also correctly loop over all maps in the json response, and not just hard-code one map.

File: common/src/main/java/com/wynntils/core/webapi/WebManager.java (hunk 4 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: Yes, that is a solution. Or we name the files after the coordinates: "-1331,33,-12,553.png" or so.

Comment (HighCrit) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: Well, if we split the larger map file into smaller maps. Or wynncraft has another expansion. We are still stuck with leftover files

File: common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java (hunk 6 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java]: The `register` call. If the ResourceLocation is created in the constructor, then how can registration of texture manager be done later?

File: common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java (hunk 7 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java:45]: Minecraft's server list stores icons at this resource location, which means that this location should be checked for an already existing icon.

File: common/src/main/java/com/wynntils/core/webapi/WebManager.java (hunk 4 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: Leftover files are really only a problem when files are renamed around, which should be rather infrequent. I feel it might be better to deal with that in a separate pr.

File: common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java (hunk 5 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java:39]: Will do.

File: common/src/main/java/com/wynntils/core/webapi/WebManager.java (hunk 8 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: Will do.

File: common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java (hunk 6 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java]: There are many ways that could be done. Flag it with a boolean, for instance. The thing that bothers me with this code is actually how you hard-code string literals in the constructor. These should have been gotten from the constructor.

File: common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java (hunk 7 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java:45]: But we switched to using our own static icon instead..?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java:45]: We use our static icon if we don't have a cache of it.

Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/mc/utils/objects/ServerIcon.java:45]: This is about the original implementation.

File: common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java (hunk 6 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/core/webapi/profiles/MapProfile.java]: This should be resolved.

File: common/src/main/java/com/wynntils/core/webapi/WebManager.java (hunk 9 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/core/webapi/WebManager.java]: Should be resolved.

**Review by magicus — DISMISSED** [review]

**Review by magicus — APPROVED** [review]
